automated_packaging/common_functions.pm

```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/perl
+
+package common_functions;
+use Exporter;
+use JSON;
+
+# Export subroutines
+our @ISA = qw( Exporter );
+our @EXPORT = qw(get_and_verify_token get_sorted_prs create_release_changelog has_backport_label);
+
+# untaint environment
+local $ENV{'PATH'} = '/usr/local/bin:/usr/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin';
+
+sub get_and_verify_token {
+    unless ( exists $ENV{GITHUB_TOKEN} ) {
+        die "You must have a GITHUB_TOKEN set";
+    }
+
+    my $github_token = $ENV{GITHUB_TOKEN};
+    if ( $ENV{GITHUB_TOKEN} =~ /^(\w+)$/x ) {
+        $github_token = $1;
+    }
+    else {
+        die "Malformed GITHUB_TOKEN: $github_token";
+    }
+
+    my $cmd = "curl -sf -H 'Accept: application/vnd.github.v3.full+json' -X GET --user '$github_token:x-oauth-basic' " . 'https://api.github.com/';
+    my $result = `$cmd > /dev/null 2>&1`;
+    my $exit_code = $? >> 8;
+
+    if ( $exit_code == 22 ) {
+        die "Your token was rejected by GitHub.";
+    }
+
+    return $github_token;
+}
+
+# Get sorted PRs according to merged-at by collecting PRs up to given data (using created info)
+sub get_sorted_prs {
+    my $earliest_pr_date = @_[0];
+    my $repo_name = @_[1];
+    my %sorted_pr_hash = ();
+    my $github_token = get_and_verify_token();
+
+    my $page_number = 1;
+    $merged_date = "2100-12-12";
+    do {
+        my $prs_text = `curl -H "Accept: application/vnd.github.v3.full+json" -X GET --user "$github_token:x-oauth-basic" 'https://api.github.com/repos/citusdata/$repo_name/pulls?base=master&state=all&page=$page_number'`;
+        my @prs = @{ decode_json($prs_text) };
+
+        foreach my $pr (@prs) {
+            my %pr_hash = %$pr;
+
+            if ( defined( $pr_hash{'merged_at'} ) ) {
+
+                $merged_date = substr( $pr_hash{'merged_at'}, 0, 10 );
+
+                if ( $merged_date lt $earliest_pr_date ) {
+                    last;
```

Comment:
This logic cannot get all PRs merged after earliest_pr_date. For example I tried with;

`prepare_changelog.pl 7.3.0 citus 2017-12-21`

It should return

```
https://api.github.com/repos/citusdata/citus/pulls/1946
https://api.github.com/repos/citusdata/citus/pulls/1939
https://api.github.com/repos/citusdata/citus/pulls/1943
https://api.github.com/repos/citusdata/citus/pulls/1929
https://api.github.com/repos/citusdata/citus/pulls/1923
https://api.github.com/repos/citusdata/citus/pulls/1905
https://api.github.com/repos/citusdata/citus/pulls/1903
https://api.github.com/repos/citusdata/citus/pulls/1918
https://api.github.com/repos/citusdata/citus/pulls/1894
https://api.github.com/repos/citusdata/citus/pulls/1907
https://api.github.com/repos/citusdata/citus/pulls/1901
https://api.github.com/repos/citusdata/citus/pulls/1897
https://api.github.com/repos/citusdata/citus/pulls/1878
https://api.github.com/repos/citusdata/citus/pulls/1884
https://api.github.com/repos/citusdata/citus/pulls/1895
https://api.github.com/repos/citusdata/citus/pulls/1893
https://api.github.com/repos/citusdata/citus/pulls/1806
https://api.github.com/repos/citusdata/citus/pulls/1818
https://api.github.com/repos/citusdata/citus/pulls/1883
https://api.github.com/repos/citusdata/citus/pulls/1876
```

but it returns
```
https://api.github.com/repos/citusdata/citus/pulls/1946
https://api.github.com/repos/citusdata/citus/pulls/1939
https://api.github.com/repos/citusdata/citus/pulls/1943
https://api.github.com/repos/citusdata/citus/pulls/1929
https://api.github.com/repos/citusdata/citus/pulls/1923
https://api.github.com/repos/citusdata/citus/pulls/1905
https://api.github.com/repos/citusdata/citus/pulls/1903
https://api.github.com/repos/citusdata/citus/pulls/1918
https://api.github.com/repos/citusdata/citus/pulls/1894
https://api.github.com/repos/citusdata/citus/pulls/1907
https://api.github.com/repos/citusdata/citus/pulls/1901
https://api.github.com/repos/citusdata/citus/pulls/1897
https://api.github.com/repos/citusdata/citus/pulls/1895
https://api.github.com/repos/citusdata/citus/pulls/1893
```
specifically it breaks the loop after first PR which is merged before given date.